File: Blind150/3.Sliding Window/GO/permutation_in_string.go
// https://leetcode.com/problems/permutation-in-string/description/
func checkInclusion(s1 string, s2 string) bool {

	l1, l2 := len(s1), len(s2)

	if l2 < l1 {
		return false
	}

	var countl1, countl2 [26]int

	for _, ch := range s1 {
		// fmt.Println(ch, ch - 'a')
		countl1[ch-'a']++
	}

	for i := 0; i < l1; i++ {
		countl2[s2[i]-'a']++
	}

	if countl1 == countl2 {
		return true
	}

	for i := 0; i < (l2 - l1); i++ {

		countl2[s2[i]-'a']--

		countl2[s2[i+l1]-'a']++

		if countl2 == countl1 {
			return true
		}
	}

	return false

}

File: Blind150/3.Sliding Window/Python/permutation_in_string.py
# https://leetcode.com/problems/permutation-in-string/description/
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:

        l1, l2 = len(s1), len(s2)

        if l2 < l1:
            return False

        p_hash = sum(map(hash, s1))
        s_hash = sum(hash(s2[i]) for i in range(l1))

        if p_hash == s_hash:
            return True

        for i in range(l2-l1):
            s_hash = s_hash - hash(s2[i]) + hash(s2[i + l1])
            if s_hash == p_hash:
                return True

        return False
